### ISSUE TITLE
Throwing a `Response` object leads to 500 error

### ISSUE THREAD

**Issue by Turbo87** [samselikoff/ember-cli-mirage#1703]
The thing that I am trying to implement is an authentication check for most of the routes. I imagined I could implement it roughly like this:

```js
function checkAuth() {
  if (!isAuthenticated()) {
    throw new Response(401, {}, { message: 'Not Authenticated' });
  }
}
```

but it seems that throwing a response does not work as I would have expected and instead it is always resulting in an Internal Server Error.

@samselikoff do you think it would be useful to check if the thrown error is a `Response` instance and in that case use that instead of creating a new 500 response? if that sounds good to you I'd be happy to open a PR :)

**Comment by samselikoff** [comment]
Oh interesting. Hm. I'm guessing you know you can `return` a Response, but you want this just so you can short-circuit a route handler at any time?

We really need a middleware system in Mirage... this has come up a lot recently and will probably be added in the near term.

**Comment by Turbo87** [comment]
> but you want this just so you can short-circuit a route handler at any time

exactly, yes

> We really need a middleware system in Mirage

yeah, that would be the other potential solution for this

**Comment by samselikoff** [comment]
I think I'm a little hesitant to merge this in, because it introduces some new "magic" (exceptions in route handlers return a 500, unless they happen to throw Responses), and in general I don't like using exceptions for program logic/control flow.

I think we should eventually introduce middleware pattern to deal with this.

For now, if you wanted to reuse this logic for your route handlers, I think your best bet might be to make a HOF:

```js
this.get('/users', withAuth((schema, request) => {
});
```

You might also be able to use the application serializer's `serialize` method to perform some checks / change the response before it goes out. But I agree we need a better story here.

**Comment by Turbo87** [comment]
yeah, the `withAuth()` is pretty much exactly the workaround that I'm using right now. unfortunately though `prettier` isn't exactly happy about it and transforms it into a multi-line statement... 😅

**Comment by samselikoff** [comment]
sucks :/

I'd be happy to prioritize middleware, or even ship it behind a feature flag so we can test it out a bit

**Comment by Turbo87** [comment]
> You might also be able to use the application serializer's serialize method to perform some checks / change the response before it goes out.

I just tried this out, but it seems like the serializer isn't used at all when I throw something in the route handler. Looking at the code, the `this.serialize()` call is only done in the successful case, but not for the error case. 🤔
